<?xml version="1.0" encoding="utf-8"?>
<!-- This URDF was automatically created by SolidWorks to URDF Exporter! Originally created by Stephen Brawner (brawner@gmail.com) 
  Commit Version: 1.6.0-1-g15f4949  Build Version: 1.6.7594.29634
  For more information, please see http://wiki.ros.org/sw_urdf_exporter -->
<robot
  name="mo_hinh_robot">
  <link
    name="base_link">
    <inertial>
   <origin
     xyz="-0.000110185966013022 -3.4221336087728E-06 0.0651865978110736"
     rpy="0 0 0" />
   <mass
     value="2.96390581559924" />
   <inertia
     ixx="0.00742546634157697"
     ixy="4.49220991340701E-08"
     ixz="1.3076566436296E-05"
     iyy="0.00743730950356191"
     iyz="8.54733585689313E-07"
     izz="0.0139951764965784" />
    </inertial>
    <visual>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/base_link.STL" />
   </geometry>
   <material
     name="">
     <color
    rgba="0.776470588235294 0.756862745098039 0.737254901960784 1" />
   </material>
    </visual>
    <collision>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/base_link.STL" />
   </geometry>
    </collision>
  </link>
  <link
    name="lidar_link">
    <inertial>
   <origin
     xyz="0.00676653011666337 -0.000207827566854595 -0.0224318722671681"
     rpy="0 0 0" />
   <mass
     value="1.33076913637071" />
   <inertia
     ixx="0.000614917940454093"
     ixy="-8.01984318980471E-07"
     ixz="0.000158243025911173"
     iyy="0.000960697335617201"
     iyz="-3.09162986587554E-06"
     izz="0.00109815200027134" />
    </inertial>
    <visual>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/lidar_link.STL" />
   </geometry>
   <material
     name="">
     <color
    rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
   </material>
    </visual>
    <collision>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/lidar_link.STL" />
   </geometry>
    </collision>
  </link>
  <joint
    name="lidar_joint"
    type="fixed">
    <origin
   xyz="0 0.000115901455767099 0.110703356935491"
   rpy="0 0 0" />
    <parent
   link="base_link" />
    <child
   link="lidar_link" />
    <axis
   xyz="0 0 0" />
  </joint>
  <link
    name="imu_link">
    <inertial>
   <origin
     xyz="-0.000190068708313807 -0.000282801952511392 0.000656651591240558"
     rpy="0 0 0" />
   <mass
     value="0.000491630657107432" />
   <inertia
     ixx="2.60740156632659E-08"
     ixy="1.20559518329719E-10"
     ixz="1.03040853232581E-10"
     iyy="1.05219690062939E-08"
     iyz="8.68518895699735E-11"
     izz="3.63603146259297E-08" />
    </inertial>
    <visual>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/imu_link.STL" />
   </geometry>
   <material
     name="">
     <color
    rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
   </material>
    </visual>
    <collision>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/imu_link.STL" />
   </geometry>
    </collision>
  </link>
  <joint
    name="imu_joint"
    type="fixed">
    <origin
   xyz="-0.000130860512975111 0 0"
   rpy="0 0 0" />
    <parent
   link="base_link" />
    <child
   link="imu_link" />
    <axis
   xyz="0 0 0" />
  </joint>
  <link
    name="left_wheel">
    <inertial>
   <origin
     xyz="1.26531361333351E-06 -1.8579774514766E-06 -0.0154233151501505"
     rpy="0 0 0" />
   <mass
     value="0.220455124017994" />
   <inertia
     ixx="4.59354203353536E-05"
     ixy="-1.64874741472444E-09"
     ixz="3.11694735802205E-09"
     iyy="4.59367152538397E-05"
     iyz="-4.569390606387E-09"
     izz="8.21417212128144E-05" />
    </inertial>
    <visual>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/left_wheel.STL" />
   </geometry>
   <material
     name="">
     <color
    rgba="0.752941176470588 0.752941176470588 0.752941176470588 1" />
   </material>
    </visual>
    <collision>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/left_wheel.STL" />
   </geometry>
    </collision>
  </link>
  <joint
    name="left_wheel_joint"
    type="continuous">
    <origin
   xyz="-0.0643580948556634 0.111471490168346 0.0122233569354912"
   rpy="1.5707963267949 0 0.523598775598299" />
    <parent
   link="base_link" />
    <child
   link="left_wheel" />
    <axis
   xyz="0 0 -1" />
    <limit
   effort="10"
   velocity="10" />
    <dynamics
   damping="0.1"
   friction="0.01" />
  </joint>



  
  <link
    name="right_wheel">
    <inertial>
   <origin
     xyz="-4.8564E-07 2.1959E-06 -0.015423"
     rpy="0 0 0" />
   <mass
     value="0.22046" />
   <inertia
     ixx="4.5934E-05"
     ixy="-7.509E-10"
     ixz="-1.2001E-09"
     iyy="4.5938E-05"
     iyz="5.4004E-09"
     izz="8.2142E-05" />
    </inertial>
    <visual>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/right_wheel.STL" />
   </geometry>
   <material
     name="">
     <color
    rgba="0.75294 0.75294 0.75294 1" />
   </material>
    </visual>
    <collision>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/right_wheel.STL" />
   </geometry>
    </collision>
  </link>
  <joint
    name="right_wheel_joint"
    type="continuous">
    <origin
   xyz="-0.064358094855663 -0.111471490168347 0.0122233569354911"
   rpy="1.5707963267949 0 2.61799387799149" />
    <parent
   link="base_link" />
    <child
   link="right_wheel" />
    <axis
   xyz="0 0 -1" />
    <limit
   effort="10"
   velocity="10" />
    <dynamics
   damping="0.1"
   friction="0.01" />
  </joint>
  <link
    name="front_wheel">
    <inertial>
   <origin
     xyz="2.061E-06 -8.9769E-07 0.015423"
     rpy="0 0 0" />
   <mass
     value="0.22046" />
   <inertia
     ixx="4.5937E-05"
     ixy="-1.2981E-09"
     ixz="-5.07E-09"
     iyy="4.5935E-05"
     iyz="2.2128E-09"
     izz="8.2142E-05" />
    </inertial>
    <visual>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/front_wheel.STL" />
   </geometry>
   <material
     name="">
     <color
    rgba="0.75294 0.75294 0.75294 1" />
   </material>
    </visual>
    <collision>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/front_wheel.STL" />
   </geometry>
    </collision>
  </link>
  <joint name="front_wheel_joint" type="continuous">
  <origin xyz="0.128716189711327 0 0.0122233569354896" rpy="1.5707963267949 0 1.5707963267949" />
  <parent link="base_link" />
  <child link="front_wheel" />
  <axis xyz="0 0 1" />
  <limit effort="10" velocity="10" />
  <dynamics damping="0.1" friction="0.01" />
</joint>
  <link
    name="arm1">
    <inertial>
   <origin
     xyz="-0.0037454 0.073761 -0.0078148"
     rpy="0 0 0" />
   <mass
     value="0.11467" />
   <inertia
     ixx="0.00036748"
     ixy="5.1856E-06"
     ixz="1.3665E-07"
     iyy="1.6241E-05"
     iyz="-2.9966E-06"
     izz="0.00037856" />
    </inertial>
    <visual>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/arm1.STL" />
   </geometry>
   <material
     name="">
     <color
    rgba="0.75294 0.75294 0.75294 1" />
   </material>
    </visual>
    <collision>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/arm1.STL" />
   </geometry>
    </collision>
  </link>
  <joint
    name="arm1_joint"
    type="revolute">
    <origin
   xyz="0 0.00500000000000155 0.226003356935507"
   rpy="1.5707963267949 0 0" />
    <parent
   link="base_link" />
    <child
   link="arm1" />
    <axis
   xyz="0 0 1" />
    <limit
   lower="-1.57"
   upper="1.57"
   effort="300"
   velocity="3" />
  </joint>
  <link
    name="arm2">
    <inertial>
   <origin
     xyz="-0.0042896 0.067997 -0.0078922"
     rpy="0 0 0" />
   <mass
     value="0.097049" />
   <inertia
     ixx="0.00032917"
     ixy="8.4112E-06"
     ixz="1.6328E-07"
     iyy="1.1935E-05"
     iyz="-2.5882E-06"
     izz="0.00033478" />
    </inertial>
    <visual>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/arm2.STL" />
   </geometry>
   <material
     name="">
     <color
    rgba="0.75294 0.75294 0.75294 1" />
   </material>
    </visual>
    <collision>
   <origin
     xyz="0 0 0"
     rpy="0 0 0" />
   <geometry>
     <mesh
    filename="package://mo_hinh_robot/meshes/arm2.STL" />
   </geometry>
    </collision>
  </link>
  <joint
    name="arm2_joint"
    type="revolute">
    <origin
   xyz="-0.00073149 0.17 -0.015"
   rpy="0 0 0" />
    <parent
   link="arm1" />
    <child
   link="arm2" />
    <axis
   xyz="0 0 1" />
    <limit
   lower="-1.57"
   upper="1.57"
   effort="200"
   velocity="3" />
  </joint>

    <gazebo reference="lidar_link">
   <sensor type="ray" name="lidar_sensor">
     <pose>0 0 0 0 0 0</pose>
     <visualize>true</visualize>
     <update_rate>30</update_rate>
     <ray>
    <scan>
      <horizontal>
        <samples>720</samples>
        <resolution>1</resolution>
        <min_angle>-1.5708</min_angle>
        <max_angle>1.5708</max_angle>
      </horizontal>
    </scan>
    <range>
      <min>0.1</min>
      <max>30.0</max>
      <resolution>0.01</resolution>
    </range>
    <noise>
      <type>gaussian</type>
      <mean>0.0</mean>
      <stddev>0.01</stddev>
    </noise>
     </ray>
     <plugin name="gazebo_ros_laser" filename="libgazebo_ros_laser.so">
    <topicName>/scan</topicName>
    <frameName>lidar_link</frameName>
     </plugin>
   </sensor>
    </gazebo>

  <!-- Vật liệu (tùy chọn, nếu muốn) -->
  <gazebo reference="base_link">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="left_wheel">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="right_wheel">
    <material>Gazebo/Green</material>
  </gazebo>
  <gazebo reference="front_wheel">
    <material>Gazebo/Red</material>
  </gazebo>


<gazebo reference="imu_link">
  <sensor name="imu_sensor" type="imu">
    <always_on>true</always_on>
    <update_rate>100</update_rate>
    <visualize>true</visualize> <!-- Bật hiển thị mũi tên -->
    <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
      <topicName>/imu</topicName>
      <bodyName>imu_link</bodyName>
      <frameName>imu_link</frameName>
      <gaussianNoise>0.01</gaussianNoise>
      <xyzOffset>0 0 0</xyzOffset>
      <rpyOffset>0 0 0</rpyOffset> <!-- Không offset góc dữ liệu -->
      <initialOrientationAsReference>false</initialOrientationAsReference>
    </plugin>
    <pose>0 0 0 0 1.5708 0</pose> <!-- Xoay 90 độ quanh trục Y để mũi tên hướng theo X -->
  </sensor>
</gazebo>


<transmission name="arm1_trans">
<type>transmission_interface/SimpleTransmission</type>
<joint name="arm1_joint">
<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
</joint>
<actuator name="arm1_motor">
<!-- <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface> -->
<mechanicalReduction>1</mechanicalReduction>
</actuator>
</transmission>

<transmission name="arm2_trans">
<type>transmission_interface/SimpleTransmission</type>
<joint name="arm2_joint">
<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
</joint>
<actuator name="arm2_motor">
<!-- <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface> -->
<mechanicalReduction>1</mechanicalReduction>
</actuator>
</transmission>

  <gazebo>
    <plugin name ="gazebo_ros_control" filename = "libgazebo_ros_control.so">
        <robotNamespace>/</robotNamespace>
    </plugin>
  </gazebo>

<!-- Transmission cho bánh xe -->
  <transmission name="left_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="left_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
    <actuator name="left_wheel_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="right_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="right_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
    <actuator name="right_wheel_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="front_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="front_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
    <actuator name="front_wheel_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

<!-- Gazebo cho left_wheel_joint -->
<gazebo reference="left_wheel_joint">
  <provideFeedback>true</provideFeedback>
  <kp>100.0</kp>
  <ki>0.1</ki>
  <kd>10.0</kd>
</gazebo>

<!-- Gazebo cho right_wheel_joint -->
<gazebo reference="right_wheel_joint">
  <provideFeedback>true</provideFeedback>
  <kp>100.0</kp>
  <ki>0.1</ki>
  <kd>10.0</kd>
</gazebo>

<!-- Gazebo cho front_wheel_joint -->
<gazebo reference="front_wheel_joint">
  <provideFeedback>true</provideFeedback>
  <kp>100.0</kp>
  <ki>0.1</ki>
  <kd>10.0</kd>
</gazebo>
</robot>